### PR TITLE
Leaves - Tiffany

### PR DESCRIPTION
**Pull request by TiffanyChio**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom method that would return a spotlight (piece of work with the highest vote count). 
Describe how you approached testing that model method. What edge cases did you come up with? | Edge cases include: it returns nil when there are no works in the database and that it returns only one instance when there is a tie in max vote count. 
What are `session` and `flash`? What is the difference between them? | Session and flash are special hashes that store information that persists longer than our standard instance variables. We use session primarily to store user information such as whether they are logged in. Session information appears to last until we clear it out or until we close the browser. However I’m told we can define this. Flash persists between redirects. We’ve used it primarily to store information we would like to display back to the user such as error messages or notifications of success. 
What was one thing that you gained more clarity on through this assignment? | Render vs redirect and flash vs flash now.
What is the Heroku URL of your deployed application? | https://tiffanys-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | There were a few things that I ended up googling such as how to redirect back to the last page I was on, how to make votes validations check for instances with the same user_id and work_id, and how/why anyone would want to put a controller filter in application controller. I think the last point was a big one. It seems like a very important concept but it wasn't covered in class. The only reason why I looked into it was because the feedback check list asked for it. Then I spent hours googling around with very little to show for it, before asking Chris the next day.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  root 'homepages#index'
+  get '/homepages', to: 'homepages#index', as: 'homepages'
+  
+  resources :works, except: [:update]
+  patch '/works/:id', to: 'works#update'
+  
+  get "/login", to: "users#login_form", as: "login"
+  post "/login", to: "users#login"
+  post "/logout", to: "users#logout", as: "logout"
+  get '/users', to: 'users#index', as: 'users'
```
Comment (beccaelenzil) [inline config/routes.rb:11]: Could you use limited resources here ?

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | this produces a 404, a redirect is better.
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | yes
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params |yes 
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Great job overall! Your code is well written and well tested. Your implementation matches the demo site very closely, and I the learning goals for this assignment were definitely met. Keep up the hard work!
